### DISCUSSION TITLE
How to populate Device Bays with a Script

### DISCUSSION THREAD

**Discussion by dremerb** [netbox-community/netbox#16163] [Q&A]
I cannot find the correct Documentation or an example script and am wondering how I would add a Device to a DeviceBay.  
When working with modules, creating a Module one can pass the target bay with the `module_bay` parameter, but there is no `device_bay` parameter in `Device` (at least in 3.7.3).

It does not appear to be correct to set `installed_device` of the target bay, since this breaks in frontend, such that `/dcim/devices/{device_id}/` page does not show a tab "Device Bays" anymore, but navigating to `/dcim/devices/{device_id}/device-bays/` correctly shows a table of bays and the devices set in the `installed_device` field.

My goal is to create fully populated blade server chassis. The device types have their "parent"/"child" flags set correctly and for every chassis I create 12 blades in a loop, but I am lacking a way to assign the blades to the correct bay.

**Comment by candlerb** [comment]
As far as I can see, the attribute is indeed `installed_device` on DeviceBay:

```
class DeviceBay(ComponentModel, TrackingModelMixin):
    """
    An empty space within a Device which can house a child device
    """
    installed_device = models.OneToOneField(
        to='dcim.Device',
        on_delete=models.SET_NULL,
        related_name='parent_bay',
        blank=True,
        null=True
    )
```

which is seen as `parent_bay` from the child Device side, e.g.

```
    def get_children(self):
        """
        Return the set of child Devices installed in DeviceBays within this Device.
        """
        return Device.objects.filter(parent_bay__device=self.pk)
```

> It does not appear to be correct to set `installed_device` of the target bay, since this breaks in frontend, such that `/dcim/devices/{device_id}/` page does not show a tab "Device Bays" anymore, but navigating to `/dcim/devices/{device_id}/device-bays/` correctly shows a table of bays and the devices set in the `installed_device` field.

This might be broken caching of the count of device bays on a device. To test this theory, try

```
select device_bay_count from dcim_device where id=N;
```

If the count is zero then it suppresses the tab in the GUI.

If that is the problem, it could be down to how you're adding the device bays to the device.

The safest option would be to put device bay templates in the device type, and then they'll be created automatically whenever you instantiate a device from the device type. If you're adding the device bays manually, then make sure you're using individual save() for each one and not using bulk_create(), because the latter bypasses the post-save hooks which update the count.

Here's a similar change I had to do: https://github.com/netbox-community/customizations/pull/101/files

Also, you *must* always call `full_clean()` before calling `save()` on any Django model (whether new or modified). If you don't, then all model validation is bypassed. Dumb, but that's just how it is.

EDIT: it's also not impossible you've found a bug of course, but in that case first I suggest you update to v3.7.8 to pick up any hidden bugfixes in that train, and then write a small standalone custom script which replicates the behaviour you're experiencing. Then it's clear if the bug is in the script itself, or deeper inside Netbox.

**Reply by dremerb** [reply]
> Also, you must always call full_clean() before calling save() on any Django model (whether new or modified). If you don't, then all model validation is bypassed. Dumb, but that's just how it is.

Something something my fault something something it's ["documented"](https://docs.netbox.dev/en/stable/customization/custom-scripts/#change-logging). (I am aware scripting is not a focus of this project :smiling_face_with_tear: )


Your answer sent me down the correct path! The issues seems to be, that i called `save()` too often.  
I was using device types with all the bay templates, as you described. After the device was created and saved, I added some tags and saved again. This did.. things..  
The tags are handled by `django-taggit`, which is not part of the Netbox Models, but the issue appears to be saving a Django model that did not receive any changes.  
Therefore the minimal example (with tagging) looks like this:
```py
chassis = Device(...)
chassis.full_clean()
chassis.save()
# device was created and saved, can add tags now. this does 
# not need validation/saving, as it is not part of the model.
chassis.tags.add("Blade Chassis")

# calling full_clean/save here breaks things.
# chassis.full_clean()
# chassis.save()

#
device_bay = DeviceBay.objects.get(device=chassis, name='bayname')
blade = Device(...)
blade.full_clean()
blade.save()
# blade created, install into chassis.
device_bay.installed_device = blade
# device_bay was changed, needs a save.
device_bay.full_clean()
device_bay.save()
```

Great hint, thanks a lot!

**Reply by candlerb** [reply]
Great. You're right, tags are a strange beast: you're basically creating separate objects in the `extras_taggeditem` table, not in the device itself. But I don't know why re-saving the original object messes them up.

FYI:

```
device_bay = DeviceBay.objects.get(device=chassis, name='bayname')
```

might simplify to something like:

```
device_bay = chassis.devicebays.get(name='bayname')
```

And if you want to be really accurate, add a `device_bay.snapshot()` immediately after that, so that the change logging works when it gets to `device_bay.save()` later.

**Comment by bitcollector1** [comment]
@candlerb 

Thanks for all you do for the NetBox community, I learn so much from you, keep up the awesome work! 

```
lso, you must always call full_clean() before calling save() on any Django model (whether new or modified). If you don't, then all model validation is bypassed. Dumb, but that's just how it is.
```

**Comment by dremerb** [comment]
Alternative workaround I found on another device :duck: 
![image](https://github.com/netbox-community/netbox/assets/9120944/3c172d44-96aa-43e2-9c2b-4bb003574250)

Do as @candlerb says, call `full_clean()` to validate your changes.

**Reply by candlerb** [reply]
Regarding the bad counters (e.g. -1 for console ports), there's a utility which should fix them:

```
/opt/netbox/venv/bin/python /opt/netbox/netbox/manage.py calculate_cached_counts
```
